cmake/configs/modes.cmake

```diff
@@ -9,104 +9,173 @@
 # Establishes build configuration modes we can use when compiling.
 
 include(CheckCCompilerFlag)
+include(CheckCXXCompilerFlag)
+include(${CMAKE_SOURCE_DIR}/cmake/helpers.cmake)
 
-set(build_modes None Debug Release)
-if("${CMAKE_C_COMPILER_ID}" STREQUAL "MSVC")
-    set(no_omit_frame_flag "/Oy-")
-else()
-    set(no_omit_frame_flag "-fno-omit-frame-pointer")
-endif()
+# Establish an internal cache variable to track our custom build modes.
+set(BUILD_MODES None Debug Release CACHE INTERNAL "")
 
-# Create an ASAN build variant
 if("${CMAKE_C_COMPILER_ID}" STREQUAL "MSVC")
-    set(asan_link_flags "/fsanitize=address")
-    set(asan_compiler_flag "/fsanitize=address")
+    set(MSVC_C_COMPILER 1)
 elseif("${CMAKE_C_COMPILER_ID}" STREQUAL "Clang" OR "${CMAKE_C_COMPILER_ID}" STREQUAL "AppleClang")
```

Comment:
Should we use a regex expression here?
`if(CMAKE_CXX_COMPILER_ID MATCHES "^(Apple)?(C|c?)lang")`

Reply:
That's a great idea, its a lot more compact. I went over and update the uses 4ac67b68de004e27df972039f9d8b12783adc783